### PR TITLE
Fix for https://github.com/google-deepmind/optax/issues/1335

### PR DESCRIPTION
**Pull request by rdyro**

### CODE REVIEW

File: optax/losses/_classification.py [hunk 1]
```diff
@@ -251,7 +251,7 @@ def softmax_cross_entropy(
     labels: One-hot encoded labels, with shape `[batch_size, num_classes]`. Each
       row represents the true class distribution for a single example.
     axis: Axis or axes along which to compute.
-    where: Elements to include in the computation.
+    where: Elements to include in the computation of shape ``[batch_size]``.
```
Comment (vroulet) [inline optax/losses/_classification.py]: Or shape ``[batch_size, num_classes]``? You let that still be viable (and said it in the comments of the issue).
Same comment for the other docstring

File: optax/losses/_classification.py [hunk 2]
```diff
@@ -388,14 +392,10 @@ def softmax_cross_entropy_with_integer_labels(
   # This is like jnp.take_along_axis(jax.nn.log_softmax(...), ...) except that
   # we avoid subtracting the normalizer from all values, just from the values
   # for the correct labels.
-  logits_max = jnp.max(
-      logits, axis, keepdims=True, where=where, initial=-jnp.inf
-  )
-  logits -= jax.lax.stop_gradient(logits_max)
   label_logits = jnp.take_along_axis(
```
Comment (vroulet) [inline optax/losses/_classification.py:397]: You need to have the where used also for label_logits

**Comment by vroulet** [comment]
Could you add a little test? #912 seems to have just skipped the test that would have helped us avoid this bug. (see lack of test_mask in that PR)

File: optax/losses/_classification.py (hunk 1 above)
Comment (rdyro) [reply optax/losses/_classification.py]: Added!

File: optax/losses/_classification.py [hunk 3]
```diff
@@ -251,7 +251,8 @@ def softmax_cross_entropy(
     labels: One-hot encoded labels, with shape `[batch_size, num_classes]`. Each
       row represents the true class distribution for a single example.
     axis: Axis or axes along which to compute.
-    where: Elements to include in the computation.
+    where: Elements to include in the computation of shape ``[batch_size]`` or
+      logits.shape.
```
Comment (emilyfertig) [inline optax/losses/_classification.py:255]: Optional: `[batch_size, num_classes]` is clearer IMO than `logits.shape`.

**Review by emilyfertig — APPROVED** [review]

File: optax/losses/_classification.py (hunk 3 above)
Comment (rdyro) [reply optax/losses/_classification.py:255]: We're technically allowing `[num_classes, batch_size]` and any other combination if `axis` is correctly specified I think

Comment (emilyfertig) [reply optax/losses/_classification.py:255]: Ah ok -- the `logits` and `labels` docstrings specify `[batch_size, num_classes]`.
